### ISSUE TITLE
OpenCL can not run multiple threads on different GPU platforms

### ISSUE THREAD

**Issue by robertoshiu** [lightvector/KataGo#228]
I have one nVidia GPU and one AMD GPU in the same machine. It only can run single thread either on nVidia OpenCL 1.2 or AMD OpenCL 2.1. Can we use multiple OpenCL devices across different platforms (nVidia, AMD, Intel) at the same time? I'm wondering because LeelaZero does.

**Comment by lightvector** [comment]
I think you should be able to do this. What happens when you follow the instructions in the config to use multiple GPUs: increasing the number of server threads and uncommenting/adding entries to set them each to the appropriate device?

**Comment by robertoshiu** [comment]
I've modified three parameters in the gtp_example.cfg:

numNNServerThreadsPerModel = 2
openclDeviceToUseThread0 = 0
openclDeviceToUseThread1 = 1

and commented one parameter
#openclDeviceToUse = 1

then run benchmark command in PowerShell:

PS C:\katago-v1.3.5-opencl-windows-x64> .\katago benchmark -model g170e-b20c256x2-s3761649408-d809581368.bin.gz -config gtp_example.cfg
2020-05-18 15:31:41+0800: Loading model and initializing benchmark...
2020-05-18 15:31:41+0800: nnRandSeed0 = 17213472869923450201
2020-05-18 15:31:41+0800: After dedups: nnModelFile0 = g170e-b20c256x2-s3761649408-d809581368.bin.gz useFP16 auto useNHWC auto
2020-05-18 15:31:42+0800: Found OpenCL Platform 0: NVIDIA CUDA (NVIDIA Corporation) (OpenCL 1.2 CUDA 10.2.95)
2020-05-18 15:31:42+0800: Found 1 device(s) on platform 0 with type CPU or GPU or Accelerator
2020-05-18 15:31:42+0800: Found OpenCL Platform 1: AMD Accelerated Parallel Processing (Advanced Micro Devices, Inc.) (OpenCL 2.1 AMD-APP (2841.19))
2020-05-18 15:31:42+0800: Found 1 device(s) on platform 1 with type CPU or GPU or Accelerator
2020-05-18 15:31:42+0800: Found OpenCL Device 0: GeForce GTX 1070 (NVIDIA Corporation) (score 11000102)
2020-05-18 15:31:42+0800: Found OpenCL Device 1: Ellesmere (Advanced Micro Devices, Inc.) (score 11000200)
PS C:\katago-v1.3.5-opencl-windows-x64>


It seems two OpenCL devices are found on two different platforms but multi-threads started unsuccessfully.


The single thread will be fine if using this parameter:
openclDeviceToUse = 1

2020-05-18 15:26:31+0800: Using OpenCL Device 0: GeForce GTX 1070 (NVIDIA Corporation) OpenCL 1.2 CUDA
2020-05-18 15:26:31+0800: Loaded tuning parameters from: C:\katago-v1.3.5-opencl-windows-x64/KataGoData/opencltuning/tune6_gpuGeForceGTX1070_x19_y19_c256_mv8.txt
2020-05-18 15:26:31+0800: OpenCL backend: Model version 8
2020-05-18 15:26:31+0800: OpenCL backend: Model name: g170-b20c256x2-s3761649408-d809581368

2020-05-18 15:26:35+0800: Loaded config gtp_example.cfg
2020-05-18 15:26:35+0800: Loaded model g170e-b20c256x2-s3761649408-d809581368.bin.gz


Did I miss something?

**Comment by lightvector** [comment]
And what is the error message in the first case, where you tried to use both devices?

**Comment by lightvector** [comment]
Oh I see, it exited for you without any error message? That makes things tricky.

Did you test that it works with *both* OpenCL devices, individually, such that in the config it says it works both when it's using Geforce and when it's using AMD?

**Comment by robertoshiu** [comment]
> Did you test that it works with _both_ OpenCL devices, individually, such that in the config it says it works both when it's using Geforce and when it's using AMD?


Yes, nVidia and AMD are both working well individually as I told in the beginning. 


2020-05-19 20:27:33+0800: **Using** OpenCL Device 1: Ellesmere (Advanced Micro Devices, Inc.) OpenCL 2.0 AMD-APP (2841.19)
2020-05-19 20:27:33+0800: Loaded tuning parameters from: C:\katago-v1.3.5-opencl-windows-x64/KataGoData/opencltuning/tune6_gpuEllesmere_x19_y19_c256_mv8.txt
2020-05-19 20:27:37+0800: OpenCL backend: Model version 8
2020-05-19 20:27:37+0800: OpenCL backend: Model name: g170-b20c256x2-s3761649408-d809581368

2020-05-19 20:27:40+0800: Loaded config gtp_example.cfg
2020-05-19 20:27:40+0800: Loaded model g170e-b20c256x2-s3761649408-d809581368.bin.gz

**Comment by lightvector** [comment]
Thanks for reporting!

I think I've found what the issue is. In theory there should have been a clear error message on your end that would have made it obvious for me where to look for the problem, but unfortunately it seems many GPU vendors have their OpenCL crash and kill the program directly on some errors instead of raising the error in a way that could be caught and printed out.

I think I found the issue (or *an* issue) that would cause this though - it's relating to illegal reuse of an internal opencl "context" across different platforms. There will be a fix next release.
